aws-signer/src/main/scala/http4s/aws/middleware/AwsSigner.scala

```diff
@@ -0,0 +1,134 @@
+// Based on https://github.com/http4s/contrib/blob/master/aws/src/main/scala/org/http4s/contrib/aws/AwsSigner.scala
+
+import java.util.Date
+
+import cats.data.Kleisli
+import cats.effect.Effect
+import cats.implicits._
+import fs2.Stream
+import org.http4s.client.Client
+import org.http4s.{Header, Request}
+import scodec.bits.ByteVector
+
+object AwsSigner {
+  def apply[F[_]: Effect](id: String,
+                          secret: String,
+                          zone: String,
+                          service: String)(client: Client[F]): Client[F] = {
+    val signer = new AwsSigner(Key(id, secret), zone, service)
+    client.copy(open = Kleisli { req =>
+      signer[F](req, Effect[F].delay(new Date)).flatMap(client.open(_))
+    })
+  }
+
+  case class Key(id: String, secret: String) {
+    def bytes =
+      ByteVector
+        .fromBase64(secret)
+        .getOrElse(throw new Exception(s"'$secret' not base64"))
+  }
+}
+
+/** Implements AWS request signing v4 as an http4s middleware.
+  * Signing described here: https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
+  */
+class AwsSigner(key: AwsSigner.Key, zone: String, service: String) {
+  val Method = "AWS4-HMAC-SHA256"
+  val Charset = java.nio.charset.Charset.forName("UTF-8")
+
+  private def dateFormat(s: String) = {
+    val f = new java.text.SimpleDateFormat(s)
+    f.setTimeZone(java.util.TimeZone.getTimeZone("UTC"))
+    f
+  }
+
+  val fullDateFormat = dateFormat("YYYYMMdd'T'HHmmss'Z'")
+  val shortDateFormat = dateFormat("YYYYMMdd")
+
+  def hash(bytes: ByteVector) = {
+    val digest = java.security.MessageDigest.getInstance("SHA-256")
+    bytes.grouped(1024 * 16) foreach { chunk =>
+      digest.update(chunk.toByteBuffer)
+    }
+
+    ByteVector(digest.digest)
+  }
+
+  def bytes(s: String) = ByteVector(s.getBytes(Charset))
+
+  def hmac(key: ByteVector, data: ByteVector) = {
+    val algo = "HmacSHA256"
+    val hmac = javax.crypto.Mac.getInstance(algo)
+
+    hmac.init(new javax.crypto.spec.SecretKeySpec(key.toArray, algo))
+    ByteVector(hmac.doFinal(data.toArray))
+  }
+
+  def sign(string: String, date: java.util.Date) = {
+    val kSecret = bytes(s"AWS4${key.secret}")
+    val kDate = hmac(kSecret, bytes(shortDateFormat.format(date)))
+    val kRegion = hmac(kDate, bytes(zone))
+    val kService = hmac(kRegion, bytes(service))
+    val kSigning = hmac(kService, bytes("aws4_request"))
+    hmac(kSigning, bytes(string))
+  }
```

Comment:
I think several of the things above this point don't need to be public, which gives us more latitude to evolve the implementation without breaking the API.

Reply:
I think everything above this can be private.